pkg/gardenlet/controller/federatedseed/extensions/shootstate_control.go

```diff
@@ -114,20 +143,50 @@ func shouldSkipExtensionObjectSync(extensionObject extensionsv1alpha1.Object, ki
 	return false
 }
 
-func updateShootStateExtensionState(extensionState *runtime.RawExtension, shootState *gardencorev1alpha1.ShootState, kind string, name string, purpose *string) error {
+func getShootStateExtensionState(shootState *gardencorev1alpha1.ShootState, kind string, name, purpose *string) (*runtime.RawExtension, []gardencorev1alpha1.NamedResourceReference) {
+	list := gardencorev1alpha1helper.ExtensionResourceStateList(shootState.Spec.Extensions)
+	s := list.Get(kind, name, purpose)
+	if s != nil {
+		return s.State, s.Resources
+	}
+	return nil, nil
+}
+
+func updateShootStateExtensionState(shootState *gardencorev1alpha1.ShootState, kind string, name, purpose *string, state *runtime.RawExtension, resources []gardencorev1alpha1.NamedResourceReference) {
 	list := gardencorev1alpha1helper.ExtensionResourceStateList(shootState.Spec.Extensions)
-	if extensionState == nil {
-		list.Delete(kind, &name, purpose)
-		shootState.Spec.Extensions = list
-		return nil
+	if state != nil || resources != nil {
+		list.Upsert(&gardencorev1alpha1.ExtensionResourceState{
+			Kind:      kind,
+			Name:      name,
+			Purpose:   purpose,
+			State:     state,
+			Resources: resources,
+		})
+	} else {
+		list.Delete(kind, name, purpose)
 	}
-	list.Upsert(&gardencorev1alpha1.ExtensionResourceState{
-		Kind:    kind,
-		Name:    &name,
-		Purpose: purpose,
-		State:   *extensionState,
-	})
 	shootState.Spec.Extensions = list
+}
+
+func updateShootStateResourceData(shootState *gardencorev1alpha1.ShootState, ref *autoscalingv1.CrossVersionObjectReference, data *runtime.RawExtension) {
+	list := gardencorev1alpha1helper.ResourceDataList(shootState.Spec.Resources)
+	if data != nil {
+		list.Upsert(&gardencorev1alpha1.ResourceData{
+			CrossVersionObjectReference: *ref,
+			Data:                        *data,
+		})
+	} else {
+		list.Delete(ref)
+	}
+	shootState.Spec.Resources = list
+}
+
+func getResourceByName(resources []gardencorev1alpha1.NamedResourceReference, name string) *gardencorev1alpha1.NamedResourceReference {
```

Comment:
This could be put into a general API helper (`pkg/apis/core/v1beta1/helper`) as it'll probably be needed more often in the future, right?

Reply:
Done